storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java

```diff
@@ -646,71 +616,70 @@ public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffs
-        } catch (SchemaException e) {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+        short version;
+        ProducerSnapshot producerSnapshot;
+        try {
+            version = byteBuffer.getShort();
+            producerSnapshot = new ProducerSnapshot(new ByteBufferAccessor(byteBuffer), version);
+        } catch (Exception e) {
             throw new CorruptSnapshotException("Snapshot failed schema validation: " + e.getMessage());
         }
+
+        if (version < ProducerSnapshot.LOWEST_SUPPORTED_VERSION || version > ProducerSnapshot.HIGHEST_SUPPORTED_VERSION)
+            throw new CorruptSnapshotException("Snapshot contained an unknown file version " + version);
+
+        long crc = producerSnapshot.crc();
+        long computedCrc = Crc32C.compute(buffer, PRODUCER_ENTRIES_OFFSET, buffer.length - PRODUCER_ENTRIES_OFFSET);
+        if (crc != computedCrc)
+            throw new CorruptSnapshotException("Snapshot is corrupt (CRC is no longer valid). Stored crc: " + crc
+                    + ". Computed crc: " + computedCrc);
+
+        List<ProducerSnapshot.ProducerEntry> producerEntries = producerSnapshot.producerEntries();
+        List<ProducerStateEntry> entries = new ArrayList<>(producerEntries.size());
+        for (ProducerSnapshot.ProducerEntry producerEntry : producerEntries) {
+            long producerId = producerEntry.producerId();
+            short producerEpoch = producerEntry.epoch();
+            int lastSequence = producerEntry.lastSequence();
+            long lastOffset = producerEntry.lastOffset();
+            long timestamp = producerEntry.timestamp();
+            int offsetDelta = producerEntry.offsetDelta();
+            int coordinatorEpoch = producerEntry.coordinatorEpoch();
+            long currentTxnFirstOffset = producerEntry.currentTxnFirstOffset();
+
+            OptionalLong currentTxnFirstOffsetVal = currentTxnFirstOffset >= 0 ? OptionalLong.of(currentTxnFirstOffset) : OptionalLong.empty();
+            Optional<BatchMetadata> batchMetadata =
+                    (lastOffset >= 0) ? Optional.of(new BatchMetadata(lastSequence, lastOffset, offsetDelta, timestamp)) : Optional.empty();
+            entries.add(new ProducerStateEntry(producerId, producerEpoch, coordinatorEpoch, timestamp, currentTxnFirstOffsetVal, batchMetadata));
+        }
+
+        return entries;
     }
 
     // visible for testing
     public static void writeSnapshot(File file, Map<Long, ProducerStateEntry> entries, boolean sync) throws IOException {
-        Struct struct = new Struct(PID_SNAPSHOT_MAP_SCHEMA);
-        struct.set(VERSION_FIELD, PRODUCER_SNAPSHOT_VERSION);
-        struct.set(CRC_FIELD, 0L); // we'll fill this after writing the entries
-        Struct[] structEntries = new Struct[entries.size()];
-        int i = 0;
+        ProducerSnapshot producerSnapshot = new ProducerSnapshot();
+        producerSnapshot.setCrc(0L); // we'll fill this after writing the entries
```

Comment:
The previous impl is writing data into `ByteBuffer`, so it has to put a temporary data as placeholder. We don't need this workaround now, as we are using our powerful serialization, right?

Reply:
Pardon me, did you mean we need to abandon the Crc field or something else ?

Reply:
Sorry I look too quick... I believe you mean there is no need to write the 0L here. We can directly write the crc32 value in setter

Reply:
Addressed all comments in the latest commits, thank you !